dev/bots/prepare_package.dart

```diff
@@ -426,6 +428,104 @@ class ArchiveCreator {
+            }
+            final dynamic versions =  decodedPackageListing['versions'];
+            if (versions is! List) {
+              throw const FormatException('.versions should be a list');
+            }
+            final Map<String, dynamic> versionDescription = versions.firstWhere(
+              (dynamic description) {
+                if (description is! Map) {
+                  throw const FormatException('.versions elements should be maps');
+                }
+                return description['version'] == version;
+              },
+              orElse: () => throw FormatException('Could not find $name-$version in package listing')
+            ) as Map<String, dynamic>;
+            final dynamic downloadUrl = versionDescription['archive_url'];
+            if (downloadUrl is! String) {
+              throw const FormatException('archive_url should be a string');
+            }
+            final dynamic archiveSha256 = versionDescription['archive_sha256'];
+            if (archiveSha256 is! String) {
+              throw const FormatException('archive_sha256 should be a string');
+            }
+            final http.Request request = http.Request('get', Uri.parse(downloadUrl));
+            final http.StreamedResponse response = await client.send(request);
+            if (response.statusCode != 200) {
+              throw Exception('Downloading ${request.url} failed. Status code ${response.statusCode}.');
+            }
+            final File archiveFile = File(
+              path.join(preloadCache.path, '$name-$version.tar.gz'),
+            );
+            await response.stream.pipe(archiveFile.openWrite());
+            final Stream<List<int>> archiveStream = archiveFile.openRead();
+            final Digest r = await sha256.bind(archiveStream).first;
+            if (hex.encode(r.bytes) != archiveSha256) {
+              throw Exception('Hash mismatch of downloaded archive');
+            }
+          } on Exception catch (e) {
+            stderr.write('Failed downloading $name-$version. $e\n');
+            if (retries > 0) {
+              stderr.write('Retrying download of $name-$version...');
+              // Retry.
+              continue;
+            } else {
+              rethrow;
+            }
+          }
+          break;
+        }
+      });
+    }
+    final Map<String, dynamic> cacheDescription =
+        json.decode(await _runFlutter(<String>['pub', 'cache', 'list'])) as Map<String, dynamic>;
+    final Map<String, dynamic> packages = cacheDescription['packages'] as Map<String, dynamic>;
+    final List<Future<void>> downloads = <Future<void>>[];
+    for (final MapEntry<String, dynamic> package in packages.entries) {
+      final String name = package.key;
+      final Map<String, dynamic> versions = package.value as Map<String, dynamic>;
+      for (final String version in versions.keys) {
+        downloads.add(fetchPackageArchive(name, version));
```

Comment:
If the pub cache on CI has multiple versions of the same package, will they all be included in the flutter.zip?

Reply:
They should, as we iterate all versions for each package, and we store them as `$name-$version.tar.gz`.

Reply:
Is that the same as the old behavior, or will the release .zip now contain many versions of the same package that aren't needed. Remember that the flutter/flutter pubspec.yaml files pin all packages to a single version.

Reply:
We should have exactly the same set of package-versions as before.

The difference is that they are distributed as tar.gz files instead of inside a prepared cache.

We do the same steps to "warm up" a cache (by running pub get for some set of packages) and then, instead of folding that cache inside the zip, we iterate it and download the archive for each package.

Reply:
> DBC
> 
> Did we manually inspect the results of `prepare_package.dart` to ensure that it has the same size and rough contents?

I'm working on this currently, will update.

Reply:
I ran prepare_package.dart before and after this change on the same beta commit:

`dart prepare_package.dart --branch=beta --no-publish --revision=1a0a03a41d6613c10bab23508dee456f184b181b --temp_dir=$HOME/git/tmp/prepare_package`

And verified that the set of packages are the same, as well as the versions. The only difference is that after this change the individual packages are stored as tarballs.